pyomo/contrib/gdpopt/loa.py

```diff
@@ -0,0 +1,281 @@
+    """The GDPopt (Generalized Disjunctive Programming optimizer) logic-based
+    outer approximation (LOA) solver.
+
+    Accepts models that can include nonlinear, continuous variables and
+    constraints, as well as logical conditions. For nonconvex problems, LOA
+    may not report rigorous lower/upper bounds.
+    """
+    CONFIG = _GDPoptAlgorithm.CONFIG()
+    _add_OA_configs(CONFIG)
+    _add_mip_solver_configs(CONFIG)
+    _add_nlp_solver_configs(CONFIG, default_solver='ipopt')
+    _add_tolerance_configs(CONFIG)
+
+    algorithm = 'LOA'
+
+    def _log_citation(self, config):
+        config.logger.info("\n" + """- LOA algorithm:
+        Türkay, M; Grossmann, IE.
+        Logic-based MINLP algorithms for the optimal synthesis of process
+        networks. Comp. and Chem. Eng. 1996, 20(8), 959–978.
+        DOI: 10.1016/0098-1354(95)00219-7.
+        """.strip())
+
+    def solve(self, model, **kwds):
+        """Solve the model.
+
+        Args:
+            model (Block): a Pyomo model or block to be solved
+
+        """
+        return super().solve(model, **kwds)
+
+    def _solve_gdp(self, original_model, config):
+        logger = config.logger
+
+        # We'll need these to get dual info after solving subproblems
+        add_constraint_list(self.original_util_block)
+
+        (master_util_block,
+         subproblem_util_block) = _get_master_and_subproblem(self, config)
+
+        master = master_util_block.model()
+        subproblem = subproblem_util_block.model()
+
+        original_obj = self._setup_augmented_Lagrangian_objective(
+            master_util_block)
+
+        self._log_header(logger)
+
+        # main loop
+        while self.iteration < config.iterlim:
+            self.iteration += 1
+
+            # solve linear master problem
+            with time_code(self.timing, 'mip'):
+                oa_obj = self._update_augmented_Lagrangian_objective(
+                    master_util_block, original_obj, config.OA_penalty_factor)
+                mip_feasible = solve_MILP_master_problem(master_util_block,
+                                                         self, config)
+                self._update_bounds_after_master_problem_solve(
+                    mip_feasible, oa_obj, logger)
+
+            # Check termination conditions
+            if self.any_termination_criterion_met(config):
+                break
+
+            with time_code(self.timing, 'nlp'):
+                self._fix_master_soln_solve_subproblem_and_add_cuts(
+                    master_util_block, subproblem_util_block, config)
+
+            # Add integer cut
+            with time_code(self.timing, "integer cut generation"):
+                add_no_good_cut(master_util_block, config)
+
+            # Check termination conditions
+            if self.any_termination_criterion_met(config):
+                break
+
+    def _setup_augmented_Lagrangian_objective(self, master_util_block):
```

Comment:
I think this is called augmented penalty in the outer-approximation literature as the we are not moving the actual constraint to the objective, but I can see this being an augmented Lagrangian (or Lagrangean according to the comments).

Reply:
Oh, good to know, thank you. I think I'll change to match the OA literature. That seems clearer.